File: src/Application/Auth/LoginFailure.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using po_prostu_silka.Application.Members;
using po_prostu_silka.Application.Notifications;
using po_prostu_silka.Application.Persistence;
using po_prostu_silka.Domain;
using po_prostu_silka.Domain.Members;

namespace po_prostu_silka.Application.Auth;

/// <summary>
/// Why the login failure is named: S-02's blocked members need a different message from a wrong
/// password. Callers must not treat <c>invalid_credentials</c> as "no such account" - it also covers
/// a wrong password.
///
/// <c>pending_approval</c> is unreachable and has been since S-01, which let a pending member sign in
/// rather than refusing them; S-16 then removed the pending state entirely. The literal survives in
/// the SPA's LoginFailureReason union, and removing it there is churn for no gain — nothing sends
/// it.
/// </summary>
public record LoginFailure(string Reason);

File: src/app/src/app/core/auth/login-failure.ts
import { createFailureMessages } from '../http/failure-messages';
import { LoginFailureReason } from './auth.models';

/**
 * What to tell someone whose sign-in was refused.
 *
 * <h2>Non-disclosure is the whole design of this table</h2>
 *
 * `invalid_credentials` covers a wrong password AND an address with no account, because the API
 * deliberately refuses to distinguish them. Saying "nie ma takiego konta" would hand an anonymous
 * caller a way to enumerate which e-mail addresses are members here. That is also why the login
 * screen is permanently a BANNER screen and never puts a refusal on a field: a message under the
 * e-mail box says which half was wrong just as loudly as the words would.
 * `login.spec.ts` pins it; AGENTS.md states the rule.
 */
const MESSAGES: Record<LoginFailureReason, string> = {
  invalid_credentials: 'Nieprawidłowy e-mail lub hasło.',
  blocked: 'Twoje konto zostało zablokowane. Skontaktuj się z obsługą siłowni.',
  // Declared but never produced: S-16 removed approval, so nothing creates a pending account. Kept
  // because the API still declares the literal (AuthEndpoints.LoginFailure) and the union mirrors
  // the API. It gets a sentence like every other reason — the table's rule is exhaustiveness, not
  // reachability, and an unreachable reason with no message is how a fallback ships to production.
  pending_approval: 'Twoje konto oczekuje na aktywację. Skontaktuj się z obsługą siłowni.',
};

/** The fallback for a refusal with no reason, or one this build does not know. */
const UNKNOWN = 'Nie udało się zalogować. Spróbuj ponownie za chwilę.';

/** The message for a sign-in refusal. See `core/http/failure-messages.ts` for the mechanics. */
export const loginFailureMessage = createFailureMessages(MESSAGES, UNKNOWN);
